### PR TITLE
feat: LightGlue-ONNX

### PR DESCRIPTION
**Pull request by fabio-sim**
#### Changes

This PR adds a wrapper class for loading and running LightGlue-ONNX models via ONNXRuntime.

Re: https://github.com/fabio-sim/LightGlue-ONNX/issues/40

`onnxruntime-gpu>=1.16` is required as a dependency to instantiate the new `OnnxLightGlue` class. (Importing it will still work without installing).
related to #2559 

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

#### Example Usage
Sample images can be found [here](https://github.com/fabio-sim/LightGlue-ONNX/tree/main/assets).
```python
import torch

from kornia.feature import DISK, OnnxLightGlue
from kornia.io import ImageLoadType, load_image

device = torch.device("cuda")

img0 = load_image("sacre_coeur1.jpg", ImageLoadType.RGB32, device=device)[None]
img1 = load_image("sacre_coeur2.jpg", ImageLoadType.RGB32, device=device)[None]

extractor = DISK.from_pretrained("depth", device=device).eval().to(device)

data = {}
with torch.no_grad():
    for key, img in [("image0", img0), ("image1", img1)]:
        features = extractor(img, n=None, window_size=5, score_threshold=0.0, pad_if_not_divisible=True)
        data[key] = {
            "image": img,
            "keypoints": features[0].keypoints[None],
            "keypoint_scores": features[0].detection_scores[None],
            "descriptors": features[0].descriptors[None],
        }

matcher = OnnxLightGlue(weights="disk_fp16", device=device)
result = matcher(data)
print(result)
```

### CODE REVIEW

**Comment by fabio-sim** [comment]
@ducha-aiki I believe some tests are failing as `onnxruntime-gpu` is not available on MacOS. Should it still be included as a dependency in `requirements-dev.txt`?

File: kornia/feature/lightglue_onnx/utils/download.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        model_dir = os.path.join(hub_dir, 'checkpoints')
+
+    try:
+        os.makedirs(model_dir)
+    except OSError as e:
+        if e.errno == errno.EEXIST:
+            # Directory already exists, ignore.
+            pass
+        else:
+            # Unexpected OSError, re-raise.
```
Comment (johnnv1) [inline kornia/feature/lightglue_onnx/utils/download.py:55]:
```suggestion
    os.makedirs(model_dir, exist_ok=True)
```

File: test/feature/test_lightglue_onnx.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            }
+        )
+
+        assert "matches" in outputs
+        assert "scores" in outputs
+
+    @pytest.mark.slow
+    def test_exception(self):
+        with pytest.raises(RuntimeError):
+            OnnxLightGlue(device="invalid device")
```
Comment (johnnv1) [inline test/feature/test_lightglue_onnx.py:36]: [can you add assert to error messages?](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.md#coding-standards)
```suggestion
        with pytest.raises(RuntimeError) as errinf:
            OnnxLightGlue(device="invalid device")
        assert 'you error message here' in str(errinf)
```

File: kornia/feature/lightglue_onnx/utils/keypoints.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from __future__ import annotations
+
+import torch
+
+from kornia.core import Tensor, tensor
+
+
+def normalize_keypoints(kpts: Tensor, size: Tensor) -> Tensor:
```
Comment (johnnv1) [inline kornia/feature/lightglue_onnx/utils/keypoints.py:8]: can you add a simple test for it?

File: kornia/feature/lightglue_onnx/lightglue.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
+from __future__ import annotations
+
+from typing import ClassVar
+
+import numpy as np
```
Comment (johnnv1) [inline kornia/feature/lightglue_onnx/lightglue.py:5]: this will needs np as dependency -- cc @edgarriba @ducha-aiki

Comment (fabio-sim) [reply kornia/feature/lightglue_onnx/lightglue.py]: Thanks for catching that! I've added a try:except import for numpy too.

**Review by ducha-aiki — APPROVED** [review]
LGTM

**Comment by ducha-aiki** [comment]
> @ducha-aiki I believe some tests are failing as `onnxruntime-gpu` is not available on MacOS. Should it still be included as a dependency in `requirements-dev.txt`?

Yes, just skip the test on the macOS then

File: requirements/requirements-dev.txt [hunk 5]
```diff
@@ -3,6 +3,7 @@ kornia-rs==0.0.8
 mypy[reports]
 numpy
 onnx
+onnxruntime-gpu>=1.16
```
Comment (johnnv1) [inline requirements/requirements-dev.txt:6]: note sure if this is the right marker https://peps.python.org/pep-0508/#environment-markers
```suggestion
onnxruntime-gpu>=1.16  # sys_platform != darwin
```

File: kornia/feature/lightglue_onnx/lightglue.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+        self.session.run_with_iobinding(binding)
+
+        matches, mscores = binding.get_outputs()
+
+        # TODO: The following is an unnecessary copy. Replace with a better solution when torch supports
+        # constructing a tensor from a data pointer, or when ORT supports converting to torch tensor.
+        # https://github.com/microsoft/onnxruntime/issues/15963
+        outputs = {
+            "matches": torch.from_numpy(matches.numpy()).to(self.device),
```
Comment (edgarriba) [inline kornia/feature/lightglue_onnx/lightglue.py:151]: is `matches` returned as numpy array ? you can use `dlpack`

```python
"matches": torch.from_dlpack(matches.numpy()).to(self.device)
```

Comment (fabio-sim) [reply kornia/feature/lightglue_onnx/lightglue.py]: Thanks! I've changed it to use `dlpack`. `matches` is actually of the `OrtValue` class.

**Review by johnnv1 — COMMENTED** [review]
i'm getting an error on `TestOnnxLightGlue.test_forward`, someone can check locally too? @edgarriba @ducha-aiki  -- the CI will not run this test, since it's just for cuda devices

```
E       RuntimeError: Error when binding input: There's no data transfer registered for copying tensors from Device:[DeviceType:1 MemoryType:0 DeviceId:0] to Device:[DeviceType:0 MemoryType:0 DeviceId:0]
```

**Comment by fabio-sim** [comment]
> i'm getting an error on `TestOnnxLightGlue.test_forward`, someone can check locally too? @edgarriba @ducha-aiki -- the CI will not run this test, since it's just for cuda devices
> 
> ```
> E       RuntimeError: Error when binding input: There's no data transfer registered for copying tensors from Device:[DeviceType:1 MemoryType:0 DeviceId:0] to Device:[DeviceType:0 MemoryType:0 DeviceId:0]
> ```

Hi, @johnnv1, I am able to reproduce this issue if I have only `onnxruntime` installed. Can you check if you have installed only `onnxruntime-gpu`?

**Comment by johnnv1** [comment]
> Hi, @johnnv1, I am able to reproduce this issue if I have only `onnxruntime` installed. Can you check if you have installed only `onnxruntime-gpu`?

i see, I think I had installed both libraries, within just  `onnxruntime-gpu` it's fine 😄

**Pull request merged**